File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();
        for (String str : strs) {
            String key = sort(str);
            if (map.containsKey(key)) {
                map.get(key).add(str);
            } else {
                List<String> list = new ArrayList<>() {{
                    add(str);
                }};
                map.put(key, list);
            }
        }
        return map.values().stream().collect(Collectors.toList());
    }
    
    public String sort(String s) {
        char[] chars = s.toCharArray();
        Arrays.sort(chars);
        return new String(chars);
    }
}

File: 0049-group-anagrams/0049-group-anagrams.kt
class Solution {
fun groupAnagrams(strs: Array<String>): List<List<String>> {
    val map = mutableMapOf<String, MutableList<String>>()
    strs.forEach {
        val key = it.toCharArray().sorted().joinToString("")
        if (!map.containsKey(key)) map[key] = mutableListOf()
        map[key]?.add(it)
    }

    return map.values.toList()
}
}